lib/matplotlib/axis.py

```diff
@@ -2267,19 +2267,15 @@ def set_inverted(self, inverted):
     def set_default_intervals(self):
         # docstring inherited
         xmin, xmax = 0., 1.
-        dataMutated = self.axes.dataLim.mutatedx()
-        viewMutated = self.axes.viewLim.mutatedx()
-        if not dataMutated or not viewMutated:
+        # only change view if dataLim has not changed:
+        if not self.axes.dataLim.mutatedx():
```

Comment:
Should this rather be `if not self.axes.dataLim.mutatedx() and not self.axes.viewLim.mutatedx():`?  (i.e. if someone explicitly called set_xlim() before, we should also just respect that).

Reply:
Yeah, if I do this, then 
```python
        munits.registry[Quantity] = quantity_converter
        fig, ax1 = plt.subplots()
        ax1.xaxis.update_units(Quantity([10], "miles"))
        fig.draw_no_output()
       assert ax1.get_xlim() == (0, 100)
```
fails.  The xlim is set somehow to (-0.06, 0.06), and the viewLim has been mutated.  However, I have no idea where that happens or why.

Reply:
The view limits are deliberately mutated by `_request_autoscale_view` in `_unit_change_handler()`. 

The issue here seems to be that we would like to have nonsingular limits due to what scale we choose, but we have a parallel nonsingular limit due to units.  

@anntzer you were probably the last to work on this.  I don't know how to make it all work together cleanly - it seems like the same concept was developed independently in units and scales, and they are clashing.

Reply:
I looked a bit into this, and I guess that the correct patch may be
```patch
diff --git i/lib/matplotlib/axes/_base.py w/lib/matplotlib/axes/_base.py
index afba7c8321..05317ac140 100644
--- i/lib/matplotlib/axes/_base.py
+++ w/lib/matplotlib/axes/_base.py
@@ -2857,8 +2857,8 @@ class _AxesBase(martist.Artist):
         if self.get_yscale() == 'log':
             y_stickies = y_stickies[y_stickies > 0]
 
-        def handle_single_axis(scale, autoscaleon, shared_axes, interval,
-                               minpos, axis, margin, stickies, set_bound):
+        def handle_single_axis(scale, autoscaleon, shared_axes, name,
+                               axis, margin, stickies, set_bound):
 
             if not (scale and autoscaleon):
                 return  # nothing to do...
@@ -2870,12 +2870,15 @@ class _AxesBase(martist.Artist):
             x_values = []
             minimum_minpos = np.inf
             for ax in shared:
-                x_values.extend(getattr(ax.dataLim, interval))
+                x_values.extend(getattr(ax.dataLim, f"interval{name}"))
                 minimum_minpos = min(minimum_minpos,
-                                     getattr(ax.dataLim, minpos))
+                                     getattr(ax.dataLim, f"minpos{name}"))
             x_values = np.extract(np.isfinite(x_values), x_values)
             if x_values.size >= 1:
                 x0, x1 = (x_values.min(), x_values.max())
+            elif getattr(self._viewLim, f"mutated{name}")():
+                # No data but explicit viewLims already set: exit.
+                return
             else:
                 x0, x1 = (-np.inf, np.inf)
             # If x0 and x1 are non finite, use the locator to figure out
@@ -2919,11 +2922,11 @@ class _AxesBase(martist.Artist):
             # End of definition of internal function 'handle_single_axis'.
 
         handle_single_axis(
-            scalex, self._autoscaleXon, self._shared_x_axes, 'intervalx',
-            'minposx', self.xaxis, self._xmargin, x_stickies, self.set_xbound)
+            scalex, self._autoscaleXon, self._shared_x_axes, 'x',
+            self.xaxis, self._xmargin, x_stickies, self.set_xbound)
         handle_single_axis(
-            scaley, self._autoscaleYon, self._shared_y_axes, 'intervaly',
-            'minposy', self.yaxis, self._ymargin, y_stickies, self.set_ybound)
+            scaley, self._autoscaleYon, self._shared_y_axes, 'y',
+            self.yaxis, self._ymargin, y_stickies, self.set_ybound)
 
     def _get_axis_list(self):
         return self.xaxis, self.yaxis
diff --git i/lib/matplotlib/axis.py w/lib/matplotlib/axis.py
index 443dee43f7..6f2ed97858 100644
--- i/lib/matplotlib/axis.py
+++ w/lib/matplotlib/axis.py
@@ -1456,8 +1456,7 @@ class Axis(martist.Artist):
         default = self.converter.default_units(data, self)
         if default is not None and self.units is None:
             self.set_units(default)
-
-        if neednew:
+        elif neednew:
             self._update_axisinfo()
         self.stale = True
         return True
@@ -2268,7 +2267,7 @@ class XAxis(Axis):
         # docstring inherited
         xmin, xmax = 0., 1.
         # only change view if dataLim has not changed:
-        if not self.axes.dataLim.mutatedx():
+        if not self.axes.dataLim.mutatedx() and not self.axes.viewLim.mutatedx():
             if self.converter is not None:
                 info = self.converter.axisinfo(self.units, self)
                 if info.default_limits is not None:
```

The problem is first that update_units called both `set_units` and `_update_axisinfo`, but `set_units` itself already calls `_update_axisinfo`, which thus ended up being called twice, and the second time `mutatedx` would be True (see (1)).  So first get rid of the duplicate call to _update_axisinfo.

The second problem, then, is that _update_axisinfo triggers unstale_viewLim (the first time), which leads to autoscaling being called, which would set the locator-defined (see (2)) default limits instead (causing (1)) (previously, this was not a problem because the second call to _update_axisinfo would undo that).  So fix autoscale_view to respect preset viewLims if there's no data and the viewLims have already been mutated (i.e., by the default_limits).

(2): Currently, both units and locators can define default limits (units do so via default_limits; locators do so via nonsingular()); it may make more sense to unify the two approaches and e.g. only support locator.nonsingular() (saying that units who want to define default limits must do so via a custom locator).  But that may be for later...

Reply:
Right, so as this stands, we lose the user setting the xlim *before* changing the units on the axes.  I actually think this is acceptable, and I don't see a way around it. What does 

```python
ax1.set_xlim(0, 10)
ax1.plot(Quantity(np.arange(3), "miles"))
```
even mean?

Reply:
ooops, sorry, cross post...

Reply:
Yes, the above seems to work with all the tests I've made and 
```python
ax1.set_xlim(0, 10)
ax1.plot(Quantity(np.arange(3), "miles"))
```